### PR TITLE
Support conditional branching in SE

### LINKED ISSUE #5308: Support conditional branching in SE
Branch only if block is reachable based on known constraints in the new Symbolic Execution engine.
```
var b = true;
if(b)
{
    // This should be explored
}
else
{
    // This should not be explored, because b had True constraint
}
```

Learning constraints on the way is in #5395

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #5308

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.cs [hunk 1]
```diff
@@ -296,5 +296,133 @@ public void Branching_VisitedSymbolicValue_IsImmutable()
...
+                "End");
+        }
+
+        [TestMethod]
+        public void Branching_BoolConstraints_ComplexCase()
+        {
+            const string code = @"
+var isTrue = true;
+var isFalse = false;
+var isChanding = boolParameter;
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.cs:402]: this variable is not needed, it is not used.

Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.cs:402]: this variable is not needed, it is not used.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.cs [hunk 2]
```diff
@@ -296,5 +296,133 @@ public void Branching_VisitedSymbolicValue_IsImmutable()
             validator.ValidateTag("GetHashCode", x => x.HasConstraint(TestConstraint.First).Should().BeFalse()); // Nobody set the constraint on that path
             validator.ValidateExitReachCount(2);    // Once for each state
         }
+
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.cs:299]: Please add other cases, with switch expression, switch statement, while loop. etc...

Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.cs:299]: Please add other cases, with switch expression, switch statement, while loop. etc...

**Review by csaba-sagi-sonarsource — CHANGES_REQUESTED** [review]
Please add some more UTs.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.cs [hunk 3]
```diff
@@ -296,5 +296,193 @@ public void Branching_VisitedSymbolicValue_IsImmutable()
...
+    Tag(""UnreachableElse"");
+}
+Tag(""End"");";
+            SETestContext.CreateCS(code).Validator.ValidateTagOrder(
+                "Reachable",
+                "End");
+        }
+
+        [TestMethod]
+        public void Branching_TrueLiteral_VisitsIfBranch_NotSupported()
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.cs:448]: This will be another PR

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.cs [hunk 4]
```diff
@@ -296,5 +296,195 @@ public void Branching_VisitedSymbolicValue_IsImmutable()
...
+    case false:
+        Tag(""False"");
+        break;
+    default:
+        Tag(""Default"");
+}
+Tag(""End"");";
+            SETestContext.CreateCS(code).Validator.ValidateTagOrder(
+                "True",
+                "False",    // This should not be here, switch statement produces BinaryOperatin with isTrue==true
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.cs:485]: Will you open an issue for this?

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.PatternMatching.cs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        {
+            const string code = @"
+var isTrue = true;
+var value = isTrue switch
+{
+    true => true,
+    false => false
+};
+Tag(""Value"", value);";
+            SETestContext.CreateCS(code).Validator.ValidateTag("Value", x => x.Should().BeNull()); // Should have BoolConstraint.True instead
```
Comment (csaba-sagi-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.PatternMatching.cs:40]: Same here, do you plane to open an issue for this?

**Review by csaba-sagi-sonarsource — APPROVED** [review]
LGTM! I left two comments about opening issues, nonetheless the PR is good to go!

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.cs (hunk 4 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.cs:485]: No, there's [MMF-2563](https://jira.sonarsource.com/browse/MMF-2563) for it

Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.cs:485]: And I might create a small PR for simple support

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.cs (hunk 3 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.cs:448]: ..more tricky than it looks: #5478 and #5480

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.cs (hunk 4 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.cs:485]: ..more tricky than it looks: https://github.com/SonarSource/sonar-dotnet/issues/5478 and https://github.com/SonarSource/sonar-dotnet/issues/5480
